File: bot.js
const { Client, GatewayIntentBits, EmbedBuilder } = require('discord.js');
const fetch = require('node-fetch');
require('dotenv').config();

// Bot configuration
const TOKEN = process.env.DISCORD_TOKEN;
let TICK_CHANNEL_ID = process.env.TICK_CHANNEL_ID ? parseInt(process.env.TICK_CHANNEL_ID) : 0;
const TICK_URL = 'http://tick.infomancer.uk/galtick.json';
const CHECK_INTERVAL = 5 * 60 * 1000; // 5 minutes in milliseconds

// Store the last known tick
let lastKnownTick = null;

// Create Discord client
const client = new Client({
    intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
    ],
});

/**
 * Fetch the latest tick from the API
 * @returns {Promise<string|null>} The last galaxy tick timestamp or null on error
 */
async function fetchTick() {
    try {
        const response = await fetch(TICK_URL, { timeout: 10000 });
        if (response.ok) {
            const data = await response.json();
            return data.lastGalaxyTick;
        } else {
            console.error(`Error fetching tick: HTTP ${response.status}`);
            return null;
        }
    } catch (error) {
        console.error(`Error fetching tick: ${error.message}`);
        return null;
    }
}

/**
 * Format the tick timestamp for display
 * @param {string} tickTimestamp - ISO 8601 timestamp
 * @returns {string} Formatted tick information
 */
function formatTick(tickTimestamp) {
    if (!tickTimestamp) {
        return "Unable to fetch tick data.";
    }

    try {
        const tickDate = new Date(tickTimestamp);
        const unixTimestamp = Math.floor(tickDate.getTime() / 1000);
        
        // Format for UTC display (24hr format)
        const formattedUTC = tickDate.toISOString().replace('T', ' ').replace(/\.\d+Z$/, ' UTC');
        
        // Use Discord's timestamp formatting for automatic timezone conversion
        // Format: <t:UNIX_TIMESTAMP:FORMAT>
        // F = Long Date/Time format (e.g., "Tuesday, October 29, 2025 6:58:09 PM")
        // Discord will automatically convert to each user's local timezone
        const discordTimestamp = `<t:${unixTimestamp}:F>`;
        
        // Relative time using Discord's format (e.g., "2 hours ago")
        const relativeTime = `<t:${unixTimestamp}:R>`;
        
        return `**Last Galaxy Tick (UTC):** ${formattedUTC}\n**Last Galaxy Tick (Your Time):** ${discordTimestamp}\n**Time ago:** ${relativeTime}`;
    } catch (error) {
        console.error(`Error formatting tick: ${error.message}`);
        return `**Last Galaxy Tick:** ${tickTimestamp}`;
    }
}

/**
 * Check for tick updates (runs periodically)
 */
async function checkTickUpdates() {
    const currentTick = await fetchTick();
    const checkTime = new Date().toISOString();
    
    console.log(`[${checkTime}] Checking for tick updates...`);
    console.log(`  Current tick from API: ${currentTick}`);
    console.log(`  Last known tick: ${lastKnownTick}`);
    
    if (!currentTick) {
        console.error('  Failed to fetch current tick from API');
        return;
    }
    
    if (currentTick !== lastKnownTick) {
        console.log(`  🚨 NEW TICK DETECTED! ${currentTick}`);
        
        // Update the stored tick
        lastKnownTick = currentTick;
        
        // Send notification to the configured channel
        if (TICK_CHANNEL_ID !== 0) {
            const channel = client.channels.cache.get(TICK_CHANNEL_ID.toString());
            if (channel) {
                console.log(`  Sending notification to channel ${TICK_CHANNEL_ID}...`);
                const embed = new EmbedBuilder()
                    .setTitle('🚨 NEW GALAXY TICK DETECTED!')
                    .setDescription(formatTick(currentTick))
                    .setColor(0x00FF00)
                    .setTimestamp()
                    .setFooter({ text: 'Data from tick.infomancer.uk' });
                
                await channel.send({ content: '@here', embeds: [embed] });
                console.log('  ✅ Notification sent successfully');
            } else {
                console.warn(`  ⚠️ Warning: Could not find channel with ID ${TICK_CHANNEL_ID}`);
            }
        } else {
            console.warn('  ⚠️ Warning: TICK_CHANNEL_ID not configured - no notification sent');
        }
    } else {
        console.log('  No new tick detected (tick unchanged)');
    }
}

// Bot event handlers
client.on('ready', async () => {
    console.log(`${client.user.tag} has connected to Discord!`);
    console.log('Monitoring tick updates every 5 minutes...');
    console.log(`Tick notification channel: ${TICK_CHANNEL_ID !== 0 ? TICK_CHANNEL_ID : 'NOT CONFIGURED'}`);
    
    // Initialize last known tick
    lastKnownTick = await fetchTick();
    console.log(`Initial tick on startup: ${lastKnownTick}`);
    
    // Start periodic check immediately, then every 5 minutes
    console.log('Starting tick monitoring...');
    checkTickUpdates(); // Run first check immediately
    setInterval(checkTickUpdates, CHECK_INTERVAL);
});

client.on('messageCreate', async (message) => {
    // Ignore bot messages
    if (message.author.bot) return;
    
    // Check for commands
    if (message.content.startsWith('!tick')) {
        await handleTickCommand(message);
    } else if (message.content.startsWith('!nexttick')) {
        await handleNextTickCommand(message);
    } else if (message.content.startsWith('!tickchannel')) {
        await handleSetTickChannel(message);
    }
});

/**
 * Handle !tick command
 */
async function handleTickCommand(message) {
    const tick = await fetchTick();
    
    if (tick) {
        const embed = new EmbedBuilder()
            .setTitle('🌌 Elite Dangerous Galaxy Tick')
            .setDescription(formatTick(tick))
            .setColor(0x0099FF)
            .setTimestamp()
            .setFooter({ text: 'Data from tick.infomancer.uk' });
        
        await message.reply({ embeds: [embed] });
    } else {
        await message.reply('❌ Unable to fetch tick data. Please try again later.');
    }
}

/**
 * Handle !nexttick command
 */
async function handleNextTickCommand(message) {
    const tick = await fetchTick();
    
    if (tick) {
        try {
            const tickDate = new Date(tick);
            const lastUnix = Math.floor(tickDate.getTime() / 1000);
            
            // Elite Dangerous ticks are approximately every 24 hours
            const estimatedNext = new Date(tickDate.getTime() + (24 * 60 * 60 * 1000));
            const nextUnix = Math.floor(estimatedNext.getTime() / 1000);
            
            // Format UTC times
            const lastTickUTC = tickDate.toISOString().replace('T', ' ').replace(/\.\d+Z$/, ' UTC');
            const nextTickUTC = estimatedNext.toISOString().replace('T', ' ').replace(/\.\d+Z$/, ' UTC');
            
            // Use Discord timestamps for user's local time
            const lastTickLocal = `<t:${lastUnix}:F>`;
            const nextTickLocal = `<t:${nextUnix}:F>`;
            const timeUntil = `<t:${nextUnix}:R>`;
            
            const embed = new EmbedBuilder()
                .setTitle('⏰ Next Tick Estimate')
                .setDescription(
                    `**Last Tick (UTC):** ${lastTickUTC}\n` +
                    `**Last Tick (Your Time):** ${lastTickLocal}\n\n` +
                    `**Estimated Next Tick (UTC):** ${nextTickUTC}\n` +
                    `**Estimated Next Tick (Your Time):** ${nextTickLocal}\n` +
                    `**Time Until:** ${timeUntil}\n\n` +
                    `*Note: This is an estimate. Actual tick time may vary.*`
                )
                .setColor(0xFFA500)
                .setTimestamp();
            
            await message.reply({ embeds: [embed] });
        } catch (error) {
            await message.reply(`❌ Error calculating next tick: ${error.message}`);
        }
    } else {
        await message.reply('❌ Unable to fetch tick data.');
    }
}

/**
 * Handle !tickchannel command (Admin only)
 */
async function handleSetTickChannel(message) {
    // Check if user has administrator permission
    if (!message.member.permissions.has('Administrator')) {
        await message.reply('❌ You need Administrator permission to use this command.');
        return;
    }
    
    TICK_CHANNEL_ID = message.channel.id;
    await message.reply(
        `✅ This channel will now receive tick notifications!\nChannel ID: ${TICK_CHANNEL_ID}\n\n` +
        `*Note: To persist this across restarts, add TICK_CHANNEL_ID=${TICK_CHANNEL_ID} to your .env file*`
    );
    console.log(`Tick channel set to: ${message.channel.name} (ID: ${TICK_CHANNEL_ID})`);
}

/**
 * Handle !tickstatus command - Show current tick monitoring status
 */
async function handleTickStatus(message) {
    const statusEmbed = new EmbedBuilder()
        .setTitle('📊 Tick Monitoring Status')
        .addFields(
            { name: '📍 Notification Channel', value: TICK_CHANNEL_ID ? `<#${TICK_CHANNEL_ID}> (ID: ${TICK_CHANNEL_ID})` : '❌ Not configured', inline: false },
            { name: '⏰ Check Interval', value: 'Every 5 minutes', inline: true },
            { name: '🔍 Last Known Tick', value: lastKnownTick ? formatTick(lastKnownTick) : '❌ Not fetched yet', inline: false },
            { name: '📡 API Status', value: 'Checking...', inline: true }
        )
        .setColor(TICK_CHANNEL_ID ? 0x00FF00 : 0xFFA500)
        .setTimestamp()
        .setFooter({ text: 'Use !tickchannel in any channel to set/update notification channel' });
    
    // Test API connectivity
    try {
        const currentTick = await fetchTick();
        if (currentTick) {
            statusEmbed.spliceFields(3, 1, { name: '📡 API Status', value: '✅ Online', inline: true });
            if (currentTick !== lastKnownTick) {
                statusEmbed.addFields({ name: '⚠️ New Tick Available', value: `A new tick is available but not yet detected. Current: ${currentTick}`, inline: false });
            }
        } else {
            statusEmbed.spliceFields(3, 1, { name: '📡 API Status', value: '❌ Error fetching data', inline: true });
        }
    } catch (error) {
        statusEmbed.spliceFields(3, 1, { name: '📡 API Status', value: `❌ Error: ${error.message}`, inline: true });
    }
    
    await message.reply({ embeds: [statusEmbed] });
}

// Handle errors
client.on('error', (error) => {
    console.error('Discord client error:', error);
});

process.on('unhandledRejection', (error) => {
    console.error('Unhandled promise rejection:', error);
});

// Login to Discord
if (!TOKEN) {
    console.error('Error: DISCORD_TOKEN not found in environment variables!');
    console.error('Please create a .env file with your Discord bot token.');
    process.exit(1);
} else {
    client.login(TOKEN);
}

File: bot.py
import discord
from discord.ext import commands, tasks
import aiohttp
import os
from datetime import datetime
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Bot configuration
TOKEN = os.getenv('DISCORD_TOKEN')
TICK_CHANNEL_ID = int(os.getenv('TICK_CHANNEL_ID', '0'))
TICK_URL = 'http://tick.infomancer.uk/galtick.json'

# Bot setup
intents = discord.Intents.default()
intents.message_content = True
bot = commands.Bot(command_prefix='!', intents=intents)

# Store the last known tick
last_known_tick = None


async def fetch_tick():
    """Fetch the latest tick from the API."""
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(TICK_URL, timeout=10) as response:
                if response.status == 200:
                    data = await response.json()
                    return data.get('lastGalaxyTick')
                else:
                    print(f"Error fetching tick: HTTP {response.status}")
                    return None
    except Exception as e:
        print(f"Error fetching tick: {e}")
        return None


def format_tick(tick_timestamp):
    """Format the tick timestamp for display."""
    if not tick_timestamp:
        return "Unable to fetch tick data."
    
    try:
        # Parse ISO 8601 timestamp
        tick_dt = datetime.fromisoformat(tick_timestamp.replace('Z', '+00:00'))
        
        # Format for display (UTC)
        formatted_utc = tick_dt.strftime('%Y-%m-%d %H:%M:%S UTC')
        
        # Convert to local time
        local_dt = tick_dt.astimezone()
        formatted_local = local_dt.strftime('%Y-%m-%d %H:%M:%S %Z')
        
        # Calculate time ago
        now = datetime.now(tick_dt.tzinfo)
        time_diff = now - tick_dt
        hours = int(time_diff.total_seconds() // 3600)
        minutes = int((time_diff.total_seconds() % 3600) // 60)
        
        time_ago = f"{hours}h {minutes}m ago" if hours > 0 else f"{minutes}m ago"
        
        return f"**Last Galaxy Tick:** {formatted_utc}\n**Last Galaxy Tick LOCAL:** {formatted_local}\n**Time ago:** {time_ago}"
    except Exception as e:
        print(f"Error formatting tick: {e}")
        return f"**Last Galaxy Tick:** {tick_timestamp}"


@bot.event
async def on_ready():
    """Called when the bot is ready."""
    print(f'{bot.user} has connected to Discord!')
    print(f'Monitoring tick updates every 5 minutes...')
    print(f'Tick notification channel: {TICK_CHANNEL_ID if TICK_CHANNEL_ID != 0 else "NOT CONFIGURED"}')
    
    # Initialize last known tick
    global last_known_tick
    last_known_tick = await fetch_tick()
    print(f'Initial tick on startup: {last_known_tick}')
    
    # Start the background task
    if not check_tick_updates.is_running():
        print('Starting tick monitoring...')
        check_tick_updates.start()


@bot.command(name='tick', help='Display the last Elite Dangerous galaxy tick')
async def tick_command(ctx):
    """Command to display the current tick."""
    tick = await fetch_tick()
    
    if tick:
        embed = discord.Embed(
            title="🌌 Elite Dangerous Galaxy Tick",
            description=format_tick(tick),
            color=discord.Color.blue(),
            timestamp=datetime.now()
        )
        embed.set_footer(text="Data from tick.infomancer.uk")
        await ctx.send(embed=embed)
    else:
        await ctx.send("❌ Unable to fetch tick data. Please try again later.")


@tasks.loop(minutes=5)
async def check_tick_updates():
    """Background task to check for tick updates every 5 minutes."""
    global last_known_tick
    
    check_time = datetime.now().isoformat()
    current_tick = await fetch_tick()
    
    print(f"[{check_time}] Checking for tick updates...")
    print(f"  Current tick from API: {current_tick}")
    print(f"  Last known tick: {last_known_tick}")
    
    if not current_tick:
        print("  Failed to fetch current tick from API")
        return
    
    if current_tick != last_known_tick:
        print(f"  🚨 NEW TICK DETECTED! {current_tick}")
        
        # Update the stored tick
        last_known_tick = current_tick
        
        # Send notification to the configured channel
        if TICK_CHANNEL_ID != 0:
            channel = bot.get_channel(TICK_CHANNEL_ID)
            if channel:
                print(f"  Sending notification to channel {TICK_CHANNEL_ID}...")
                embed = discord.Embed(
                    title="🚨 NEW GALAXY TICK DETECTED!",
                    description=format_tick(current_tick),
                    color=discord.Color.green(),
                    timestamp=datetime.now()
                )
                embed.set_footer(text="Data from tick.infomancer.uk")
                await channel.send("@here", embed=embed)
                print("  ✅ Notification sent successfully")
            else:
                print(f"  ⚠️ Warning: Could not find channel with ID {TICK_CHANNEL_ID}")
        else:
            print("  ⚠️ Warning: TICK_CHANNEL_ID not configured - no notification sent")
    else:
        print("  No new tick detected (tick unchanged)")


@check_tick_updates.before_loop
async def before_check_tick_updates():
    """Wait until the bot is ready before starting the loop."""
    await bot.wait_until_ready()


@bot.command(name='tickchannel', help='Set the channel for tick notifications (Admin only)')
@commands.has_permissions(administrator=True)
async def set_tick_channel(ctx):
    """Command to set the current channel as the tick notification channel."""
    global TICK_CHANNEL_ID
    TICK_CHANNEL_ID = ctx.channel.id
    await ctx.send(f"✅ This channel will now receive tick notifications!\nChannel ID: {TICK_CHANNEL_ID}")
    print(f"Tick channel set to: {ctx.channel.name} (ID: {TICK_CHANNEL_ID})")


@bot.command(name='nexttick', help='Estimate when the next tick might occur')
async def next_tick(ctx):
    """Command to estimate the next tick time."""
    tick = await fetch_tick()
    
    if tick:
        try:
            tick_dt = datetime.fromisoformat(tick.replace('Z', '+00:00'))
            
            # Elite Dangerous ticks are approximately every 24 hours
            # but can vary, so this is just an estimate
            from datetime import timedelta
            estimated_next = tick_dt + timedelta(hours=24)
            now = datetime.now(tick_dt.tzinfo)
            time_until = estimated_next - now
            
            hours_until = int(time_until.total_seconds() // 3600)
            minutes_until = int((time_until.total_seconds() % 3600) // 60)
            
            embed = discord.Embed(
                title="⏰ Next Tick Estimate",
                description=f"**Last Tick:** {tick_dt.strftime('%Y-%m-%d %H:%M:%S UTC')}\n"
                           f"**Estimated Next Tick:** {estimated_next.strftime('%Y-%m-%d %H:%M:%S UTC')}\n"
                           f"**Time Until:** ~{hours_until}h {minutes_until}m\n\n"
                           f"*Note: This is an estimate. Actual tick time may vary.*",
                color=discord.Color.orange(),
                timestamp=datetime.now()
            )
            await ctx.send(embed=embed)
        except Exception as e:
            await ctx.send(f"❌ Error calculating next tick: {e}")
    else:
        await ctx.send("❌ Unable to fetch tick data.")


# Run the bot
if __name__ == '__main__':
    if not TOKEN:
        print("Error: DISCORD_TOKEN not found in environment variables!")
        print("Please create a .env file with your Discord bot token.")
    else:
        bot.run(TOKEN)
